source/text/typesetter.ts

```diff
@@ -226,84 +307,192 @@ export class Typesetter {
+            return;
+        }
+
+        let offset = -width;
+        if (alignment === Label.Alignment.Center) {
+            offset *= 0.5;
+        }
+
+        /* Origin is expected to be in typesetting space (not transformed yet). */
+        for (let i = begin; i < end; ++i) {
+            vertices.vertices[i].origin[0] += offset;
+        }
+    }
 
     /**
-     * Transforms the labels extent (typesetting space to, e.g., world space or screen space).
-     * @param transform - Transformation that was applied to every vertex.
-     * @param extent - Untransformed label extent (in typesetting space) to be transformed.
-     * @returns - Transformed label extent (copy).
+     * Create and transform glyph vertices for rendering.
+     * @param label -
+     * @param vertices -
+     * @param lines -
      */
-    protected static transformExtent(transform: mat4, extent: vec2): GLfloat2 {
-        const lowerLeft = vec4.transformMat4(v4(), vec4.fromValues(0.0, 0.0, 0.0, 1.0), transform);
-        const lowerRight = vec4.transformMat4(v4(), vec4.fromValues(extent[0], 0.0, 0.0, 1.0), transform);
-        const upperLeft = vec4.transformMat4(v4(), vec4.fromValues(0.0, extent[1], 0.0, 1.0), transform);
-        return [vec4.distance(lowerRight, lowerLeft), vec4.distance(upperLeft, lowerLeft)];
+    private static transform(label: Label, vertices: GlyphVertices, lines: Array<Line>): void {
+        for (const line of lines) {
+            Typesetter.transformAlignment(line[2], label.alignment, vertices, line[0], line[1]);
+            Typesetter.transformVertex(label.staticTransform, vertices, line[0], line[1]);
+        }
     }
 
 
     /**
      * Typesets the given label, transforming the vertices in-world, ready to be rendered.
-     * @param label the label that shall be typeset
-     * @param glyphs the glyph vertices, a prepared (optionally empty) vertex storage
-     * @param begin vertex index to start the typesetting (usually 0)
+     * @param label - The label that is to be typeset.
+     * @param vertices - The glyph vertices, a prepared (optionally empty) vertex storage.
      */
-    static typeset(label: Label, glyphs?: GlyphVertices, begin?: number): GLfloat2 {
-        assert(!!label.fontFace, `expected a font face for label before typesetting`);
+    static typeset(label: Label, vertices: GlyphVertices): void {
+
+        /** @todo its  wrong to get the vertices here when we do not even know how many glyphs will be typeset. */
+
+        if (label.length === 0) {
+            return;
+        }
+
+        assert(label.fontFace !== undefined, `expected a font face for label before typesetting`);
+        const fontFace = label.fontFace!;
+
+        /* Retrieve advances, kernings, as well as line feed and delimiter indices. */
+
+        const glyphs = (index: number): Glyph => index < label.length ? fontFace.glyph(label.charCodeAt(index)) :
+            fontFace.glyph(label.ellipsis.charCodeAt(index - label.length));
```

Comment:
I don't understand the semantics behind the `index - label.length` math. The `ellipsis` is supposed to be a very short character sequence, right?